Added user account creation in Dexter VS plugin

KarolAntczak opened the pull request:
When user provides correct server address but wrong credentials and clicks 'OK', a following message appears, allowing to create new account:

![image](https://user-images.githubusercontent.com/13940703/27037891-a447695c-4f89-11e7-82c4-402dc3989d5d.png)


After clicking 'yes' a new account is creted, indicated by appropriate message:

![image](https://user-images.githubusercontent.com/13940703/27037897-a844fc7c-4f89-11e7-9c7e-56631dcafe80.png)


Also, fixed issue #142

KarolAntczak commented:
@mdopark OK, this is now performed by DexterClient (I added appropriate function) and doesn't require dexter jar. 
Also I found a little bug in server - fixed it.

mdopark commented on project/dexter-vs/Dexter/Common/Core/Client/DexterClient.cs:
```diff
@@ -79,6 +81,21 @@ public Task StoreSourceCodeCharSequence(long snapshotId, long groupId, string mo
+
+            if (!response.IsSuccessStatusCode.Equals(true))
+            {
+                Debug.WriteLine(response, "Failed to add account!");
```
The error is being ignored. It would be better to return the success / failure of account creation.

KarolAntczak replied:
Ok, I will modify it.

mdopark commented on project/dexter-vs/Dexter/Core/UI/Settings/SettingsControl.cs:
```diff
@@ -198,9 +198,42 @@ DexterInfo IDexterInfoProvider.Load()
+            }
+            else
+            {
+                DexterClient.Instance = new DexterClient(new DexterHttpClientWrapper(dexterInfoProvider), dexterInfoProvider);
```
DexterClient.Instance is a singleton instance and It is created already in  InitDexterClient() of DexterCommandPackage.
You don't need to create it again.

KarolAntczak replied:
I think it could be refactored - there is no need to have singleton client in case of our plugin at all.

mdopark replied:
Can you provide a detailed basis for your claim to DexterClient? Why can not I use the singletone instance of DexterClient and always create a new instance?

KarolAntczak replied:
Yes, of course. I have several arguments. 

1. **There is no need to use singletons**

Typical implementation of REST API client is to create separate instance of client **each time it is needed**, without using singletons. Just take a look at typical implementations available on the web: 

- [code 1](https://www.codeproject.com/Tips/497123/How-to-make-REST-requests-with-Csharp)
- [code 2](https://stackoverflow.com/questions/9620278/how-do-i-make-calls-to-a-rest-api-using-c)

Singleton instance of REST API client makes sense in several cases, e.g. when we need to maintain connection through whole application liftetime. But this is not the case. 

2. **Implementation issues**

Singleton implementation of Dexter Client is broken for several reasons. Constructor should not be publicly exposed. Instance creation should be performed in getter. Moreover, it is being used in a way which replaces previous instance:

`DexterClient.Instance = new DexterClient(new DexterHttpClientWrapper(dexterInfoProvider), dexterInfoProvider);`
				
3. **Implementation issues cont.**

Even if this singleton creation would be OK, we are still, in fact, not using singletons. Consider this code from DexterHttpClientWrapper:

        public Task<HttpResponseMessage> GetAsync(string requestUri)
        {
            HttpClient httpClient = new HttpClient();
            setRequestHeader(httpClient);
            return httpClient.GetAsync(requestUri);
        }

As you can see, each time a GetAsync is called, a new HttpClient is being created. So each time we are creating new instance, even though we are using the same instance of DexterClient. And all methods in DexterHttpClientWrapper are implemented this way. 

3. **Singletons are antipattern**

Even though Singletons were originally one of design patterns invented by Gang of Four, there are often considered antipattern nowadays. They are overused, hard to test, they violate OOP and are error-prone. 

- [source 1](http://archive.oreilly.com/cs/user/view/cs_msg/23417) 
- [source 2](https://stackoverflow.com/questions/137975/what-is-so-bad-about-singletons)

To cite one of the authors: 

> A singleton, as defined by the GoF, is a data structure which:
> 
> - Grants global access to an object, and
> - Enforces that only one instance of the object can ever exist.
> The first is generally considered a bad thing. We don't like globals. The second is a bit more subtle, but generally, there are virtually no cases where this is a reasonable restriction to enforce.
> 

So I would strongly advise to avoid them.

mdopark replied:
Thank you for the detailed explanation. I agree with your claim to singleton. 
In the future, I will modify it so that it does not use a singleton.

mdopark commented on project/dexter-vs/Dexter/Core/UI/Settings/SettingsControl.cs:
```diff
@@ -198,9 +198,42 @@ DexterInfo IDexterInfoProvider.Load()
+            {
+                DexterClient.Instance = new DexterClient(new DexterHttpClientWrapper(dexterInfoProvider), dexterInfoProvider);
+                DexterClient.Instance.AddAccount(dexterInfo.userName, dexterInfo.userPassword, false);
+                MessageBox.Show(string.Format("Created new user {0}", dexterInfo.userName), "Dexter info", MessageBoxButtons.OK, MessageBoxIcon.Information);
```
Because AddAccount() is async function, 
The message box is shown before the completion of this function.

The message box should be shown after checking the completion of AddAccount() fhunction.
And if the account creation fails, the message box might display a message related to this.

KarolAntczak replied:
Ok, I will add some error message.

KarolAntczak commented:
@mdopark  I refactored DexterClient so it is not a singleton anymore. 
I also modified its methods to return HttpResponseMessage, so it is easy to verify and handle result of the request.

mdopark commented on project/dexter-vs/Dexter/PeerReview/Core/PeerReviewTaggerProvider.cs:
```diff
@@ -22,6 +23,12 @@ public class PeerReviewTaggerProvider : ITaggerProvider
         public IClassifierAggregatorService AggregatorService;
 
+        [Import]
+        IDexterClient dexterClient;
```
@KarolAntczak Thanks for your fix.

The import attribute is being used incorrectly. This causes the PeerReviewTaggerProvider to fail to execute properly.

In the case of the AggregatorService, the import attribute works normally because it is a service exported by VS Editor. However, because dexterClient and dexterInfoProvider are our own classes, we need to provide an environment to export them in order to use import attributes.

KarolAntczak commented:
I added missing `[Export]` attributes to `DexterClient` and `SettingsStoreDexterInfoProvider` classes. 

This allows to add instances of these classes to `PeerReviewTaggerProvider` at runtime. I added unit test (PeerReviewTaggerProviderTest) which illustrates this idea.

mdopark commented:
@KarolAntczak Hi, Karol. Thanks for your fix.
Unit test runs normally, but PeerReviewTaggerProvider.CreateTagger () still does not execute. This method is executed by the VS editor when opening a document. I confirmed that this method executes normally, after deleting import related code.

And httpClient is not created every time in DexterClient, it seems to have to create the instance of DexterClient each time it calls the related method.

KarolAntczak commented:
Oh, thank you, now I see where the problem is. I will fix it soon.

KarolAntczak commented:
I fixed this issue. Now CreateTagger() executes properly when opening a document.

mdopark commented:
@KarolAntczak Thansk you.
I confirmed CreateTagger() excutes properly.
But I have an error about DexterClient. 
After merging it, I will correct this error.
